README.md
# Overview

    de-mur-rage

    1: the detention of a ship by the freighter beyond the time allowed for loading, unloading, or sailing

    2: a charge for detaining a ship, freight car, or truck

This ERC20 smart contract implementation for the EVM imposes a demurrage
on all held token balances.

The demurrage is a continuous value *decay*, subtracted from all
balances every minute.

Also. a time period is defined at contract creation time at which the
difference between held balances and the demurrage can be withdrawn to a
pre-selected address, which in turn can redistribute that token value.

In short: Everyone is taxed a little something every minute, and every
so often a decision is made on how to redistribute that tax.

## Features

- Continuous decay of all balances.

- Capture and redistribution of decayed balances.

- Per-minute decay resolution.

- Minting and burning of vouchers.

- Grant and revoke access to mint and burn vouchers.

- Voucher expiration (modifiable anytime after publishing).

- Supply cap (modifiable anytime after publishing).

- Constant gas usage across exponential calculations.

## Nomenclature

‘`Demurrage`’  
A percentage of token supply that will continuously be removed.

‘`Demurrage Period`’  
A period of time denominated in minutes after which demurraged amounts
are available for redistribution.

‘`Sink Account`’  
The intermediate beneficiary of the demurraged amount, which may or may
not redistribute value.

‘`Base balance`’  
The inflated balance of each used which is stored for bookkeeping.

# Use Case

The use-case inspiring this implementation is in the context of issuance
of a *voucher* representing a credit obligation of an *Issuer* or
*Association of Issuers*.

This voucher can be redeemed as payment for the products of the Issuer.

The Issuer is the entity legally obligated to redeem the voucher as
payment.

Introducing demurrage on this vehicle discourages *withholding* the
voucher, for example for speculative advantage.

This also encourages increased *velocity* of voucher use.

## Example

Given:

- 10 voucher holders.

- A total supply of 1000 tokens.

- Demurrage of 2% per 30 days (43200 minutes).

- Redistribution period of 30 days (43200 minutes).

If no trades are made, the resulting balances after one redistribution
period of every user would be 98 Vouchers.

The Sink Address will have a balance of 20 vouchers after the same
period.

Note that after the redistribution the total of all balances will equal
the total minted amount.

Note that all accounts holding such vouchers are effected by demurrage
(even the Sink Account, pending redistribution).

# Smart contract

## Common interfaces

The smart contract is written in solidity, compatible with 0.8.x.

It implements a number of interfaces both from the Ethereum (ERC)
standards aswell as the Community Inclusion Currency contract interface
suite.

### ERC standard interfaces

- [ERC20 - Token Standard](https://eips.ethereum.org/EIPS/eip-20)

- [ERC165 - Standard Interface
  Detection](https://eips.ethereum.org/EIPS/eip-165)

- [ERC173 - Contract Ownership
  Standard](https://eips.ethereum.org/EIPS/eip-173)

- [ERC5679 - Token Minting and Burning (as part of CIC.Minter and
  CIC.Burner)](https://eips.ethereum.org/EIPS/eip-5679)

### CIC interfaces

- [Burner](https://git.grassecon.net/cicnet/cic-contracts/src/branch/master/solidity/Burner.sol)

- [Expire](https://git.grassecon.net/cicnet/cic-contracts/src/branch/master/solidity/Expire.sol)

- [Minter](https://git.grassecon.net/cicnet/cic-contracts/src/branch/master/solidity/Minter.sol)

- [Seal](https://git.grassecon.net/cicnet/cic-contracts/src/branch/master/solidity/Seal.sol)

- [Writer](https://git.grassecon.net/cicnet/cic-contracts/src/branch/master/solidity/Writer.sol)

## Dependencies

The token contract uses the
[ADBKMath](https://github.com/abdk-consulting/abdk-libraries-solidity/blob/master/ABDKMath64x64.sol)
library to calculate exponentials.

## Permissions

The smart contract defines three levels of access.

1.  Voucher contract owner

2.  Voucher minter

3.  Voucher holder

### Contract owner

When the contract is published to the network, the signer account of the
publishing transaction will be the contract owner.

Contract ownership can be changed by the owner using the **ERC173**
standard interface.

### Minter

A minter has access to mint vouchers, and to burn vouchers from its own
balance.

Only the contract owner may mint, and may add and remove minters.
Minters may be added and removed using the **CIC Writer** interface, as
long as the `WRITER_STATE` seal is not set. See [Sealing the
contract](#seal_005fstate) for further details.

The contract owner is automatically a minter.

### Holder

Any address may hold vouchers, and transfer vouchers from their balance.

Minters and the contract owner are automatically token holders.

All token holders are subject to demurrage.

## Publishing the contract

The contract is published with the following arguments:

‘`name`’  
ERC20 voucher name

‘`symbol`’  
ERC20 voucher symbol

‘`decimals`’  
ERC20 decimal count

‘`decayLevel`’  
Level of decay per minute. See [Specifying
demurrage](#specifying_005fdemurrage) below for further details.

‘`periodMinutes`’  
Number of minutes between each time the demurraged value can be
withdrawn to the *Sink Account*. See [Withdrawing demurraged
value](#withdrawing) below for further details. The period may not be
altered.

‘`defaultSinkAddress`’  
The initial *Sink Address*. The address may be altered as long as the
`SINK_STATE` seal has not been set. See [Sealing the
contract](#seal_005fstate) for further details.

### Specifying demurrage

The *input parameter* to the contract is a 128-bit positive fixed-point
number, where the most significant 64 bits represent the integer part,
and the lower 64 bits represents the decimals part, each consecutive
lesser bit halving the value of the previous bit.

For example, The byte value `00000000 00000002 a0000000 00000000`,
representing a zero-stripped binary value of $10.101$. This translates
to the (base 10) decimal value $2.625$. The decimal part is calculated
as, from left to right: $(1 * 0.5) + (0 * 0.25) + (1 * 0.125)$.

#### Calculating the demurrage parameter

The minute granularity of the demurrage value is calculating using the
continuous decay function.

For example, for a demurrage of 2% per 30 days (43200 minutes), the
input value will be:

$(1-0.02)^(1/43200) ~ 0.99999953234484737109$

The decimal part of the fixed-point representation of this value is:

`fffff8276fb8cfff`

The input parameter becomes:

`0000000000000000ffffa957014dc7ff`

See [Tools](#tools) for additional help generating the necessary values.

Note that attempting to publish a voucher contract with no (zero)
demurrage will fail (if demurrage is not needed, use another contract).

## Using the contract

### Withdrawing demurrage

After each redistribution period, the demurraged value of that period
can be withdrawn to the currently defined *Sink Account*.

The demurrage is calculated as from the total supply of voucher at the
end of the period.

Withdrawal should happen implicitly duing normal operation of the
contract. See [Side-effects in state changes](#sideeffects).

To explicitly credit the *Sink Address* with the demurrage value after a
period has been exceeded, the `changePeriod()` (`8f1df6bc`) method can
be called.

### Setting voucher expiry

The effect of a voucher expiring is that all balances will be frozen,
and all state changes affecting token balances will be blocked.

Expiry is defined in terms of redistribution periods. For example, if
the redistribution period is 30 days, and the expity is 3, then the
voucher expires after 90 days.

The expiry takes effect immediately when the redistribution period time
has been exceeded.

When the contract is published, no expiry is set.

Expiry may be set after publishing using the `CIC.Expire` interface.

If the `EXPIRE_STATE` seal has been set, expiry may not be changed
further.

### Capping voucher supply

The effect of a voucher supply cap is that all `CIC.Minter` calls will
fail if the total supply after minting exceeds the defined supply cap.

The supply cap still allows vouchers to be minted after `CIC.Burn`
calls, provided that the previous condition holds.

To apply the supply cap, the method `setMaxSupply(uint256) (6f8b44b0)`
is used.

### Side-effects in state changes

All state changes involving voucher values implicitly execute two core
methods to ensure application of the demurrage and redistribution.

The two methods are:

`applyDemurrage() (731f237c)`  
Calculates the demurrage modifier of all balances according to the
current timestamp.

`changePeriod() (8f1df6bc)`  
If the previously executed period change does not match the current
period, the period is changed, and the *Sink Address* is credited with
the demurrage amount of the current total supply.

Both of these methods are *noop* if no demurrage or withdrawal is
pending, respectively.

Examples of state changes that execute these methods include
`ERC20.transfer(...)`, `ERC20.transferFrom(...)` and `CIC.mintTo(...)`.

### Sealing the contract

Certain mutable core parameters of the contract can be *sealed*, meaning
prevented from being modifier further.

Sealing is executed using the `CIC.Seal` interface.

The sealing of parameters is irreversible.

The sealable parameters are[^1]:

`WRITER_STATE`  
The `CIC.Writer` interface is blocked. The effect of this is that no
more changes may be made to which accounts have minter permission.

`SINK_STATE`  
After setting this seal, the *Sink Address* may not be changed.

`EXPIRY_STATE`  
Prevents future changes to the voucher expiry date[^2].

`CAP_STATE`  
Immediately prevents future voucher minting, regardless of permissions.

## Gas usage

Gas usage is constant regardless of the amount of time passed between
each execution of demurrage and redistribution period calculations.

## Caveats

A `ERC20.transferFrom(...)` following an `ERC20.approve(...)` call, when
called across period thresholds, may fail if margin to demurraged amount
is insufficient.

# Tools

When installed as a python package, `erc20-demurrage-token` installs the
`erc20-demurrage-token-publish` executable script, which can be used to
publish smart contract instances.

While the man page for the tool can be referred to for general
information of the tool usage, two argument flags warrant special
mention in the context of this documentation.


python/CHANGELOG
- 0.5.5
	* Make allowance method public
- 0.5.4
	* Add Transfer() event emission to sweep() in contract
- 0.5.3
	* Add texinfo documentation
	* Add man page for publish tool
- 0.5.2
	* Update ERC165 interface response for Expire
	* Add ExpiryChange event
- 0.5.1
	* Update ERC165 interface responses
- 0.5.0
	* Change license
- 0.4.2
	* Correct burn interface implementation
- 0.4.1
	* Eliminate solidity compiler warnings
- 0.4.0
	* Rename taxLevel to decayLevel in contract
	* Add sweep contract method to fully empty one account into another
- 0.3.6
	* Reinstate owner as minter by default
- 0.3.0
	* Smart contracts use abdk math libraries, all exponential operations are static gas cost
	* Add expiry features, after which balances are frozen and no more transfers or demurrage will occur
	* Add sealable features for supply, sink address, expiry and minters (when sealed cannot be changed)
	* Deployer script now takes demurrage amount as ppm instead of literal growth fraction
	* Retire old multi and cap contracts
	* Replace contract bitfields for redistributions with structs
- 0.2.0
	* Add token burn function
	* Fix gas leak when calculating decay on period change
	* Remove all but SingleNocap contract in make install
	* Make approve explicitly set value
	* Add increaseAllowance and decreaseAllowance methods
- 0.1.1
	* Settable demurrage steps for apply demurrage cli tool
- 0.1.0
	* Dependency upgrades
- 0.0.11
	* Apply demurrage cli tool
- 0.0.10
	* Settable sink address
- 0.0.9
	* Correct redistribution amount for SingleNocap contract
- 0.0.2
	* Move to chainlib-eth
- 0.0.1
	* Interface for redistributed and non-redistributed, with or without cap


python/erc20_demurrage_token/token.py
# standard imports
import os
import logging

# external imports
from chainlib.eth.tx import (
        TxFactory,
        TxFormat,
        )
from chainlib.hash import keccak256_string_to_hex
from chainlib.eth.contract import (
        ABIContractEncoder,
        ABIContractDecoder,
        ABIContractType,
        abi_decode_single,
        )
from chainlib.eth.constant import ZERO_ADDRESS
from chainlib.jsonrpc import JSONRPCRequest
from eth_erc20 import ERC20
from hexathon import (
        add_0x,
        strip_0x,
        )
from dexif import from_fixed

# local imports
from erc20_demurrage_token.data import data_dir
from erc20_demurrage_token.seal import SealedContract
from erc20_demurrage_token.expiry import ExpiryContract

logg = logging.getLogger(__name__)


class DemurrageRedistribution:
    
    def __init__(self, v):
        d = ABIContractDecoder()
        v = strip_0x(v)
        d.typ(ABIContractType.UINT256)
        d.typ(ABIContractType.UINT256)
        d.typ(ABIContractType.BYTES32)
        d.val(v[:64])
        d.val(v[64:128])
        d.val(v[128:192])
        r = d.decode()

        self.period = r[0]
        self.value = r[1]
        self.demurrage = from_fixed(r[2])


    def __str__(self):
        return 'period {} value {} demurrage {}'.format(self.period, self.value, self.demurrage)


class DemurrageTokenSettings:

    def __init__(self):
        self.name = None
        self.symbol = None
        self.decimals = None
        self.demurrage_level = None
        self.period_minutes = None
        self.sink_address = None


    def __str__(self):
        return 'name {} demurrage level {} period minutes {} sink address {}'.format(
                self.name,
                self.demurrage_level,
                self.period_minutes,
                self.sink_address,
                )


class DemurrageToken(ERC20, SealedContract, ExpiryContract):

    __abi = {}
    __bytecode = {}

    def constructor(self, sender_address, settings, tx_format=TxFormat.JSONRPC, version=None):
        code = self.cargs(settings.name, settings.symbol, settings.decimals, settings.demurrage_level, settings.period_minutes, settings.sink_address, version=version)
        tx = self.template(sender_address, None, use_nonce=True)
        tx = self.set_code(tx, code)
        return self.finalize(tx, tx_format)


    @staticmethod
    def cargs(name, symbol, decimals, demurrage_level, period_minutes, sink_address, version=None):
        code = DemurrageToken.bytecode()
        enc = ABIContractEncoder()
        enc.string(name)
        enc.string(symbol)
        enc.uint256(decimals)
        enc.uint256(demurrage_level)
        enc.uint256(period_minutes)
        enc.address(sink_address)
        code += enc.get()
        return code


    @staticmethod
    def gas(code=None):
        return 7000000


    @staticmethod
    def abi():
        name = 'DemurrageTokenSingleNocap'
        if DemurrageToken.__abi.get(name) == None:
            f = open(os.path.join(data_dir, name + '.json'), 'r')
            DemurrageToken.__abi[name] = json.load(f)
            f.close()
        return DemurrageToken.__abi[name]


    @staticmethod
    def bytecode(version=None):
        name = 'DemurrageTokenSingleNocap'
        if DemurrageToken.__bytecode.get(name) == None:
            f = open(os.path.join(data_dir, name + '.bin'), 'r')
            DemurrageToken.__bytecode[name] = f.read()
            f.close()
        return DemurrageToken.__bytecode[name]


    def increase_allowance(self, contract_address, sender_address, address, value, tx_format=TxFormat.JSONRPC):
        enc = ABIContractEncoder()
        enc.method('increaseAllowance')
        enc.typ(ABIContractType.ADDRESS)
        enc.typ(ABIContractType.UINT256)
        enc.address(address)
        enc.uint256(value)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx


    def decrease_allowance(self, contract_address, sender_address, address, value, tx_format=TxFormat.JSONRPC):
        enc = ABIContractEncoder()
        enc.method('decreaseAllowance')
        enc.typ(ABIContractType.ADDRESS)
        enc.typ(ABIContractType.UINT256)
        enc.address(address)
        enc.uint256(value)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx


    # backwards compatibility
    def add_minter(self, contract_address, sender_address, address, tx_format=TxFormat.JSONRPC):
        return self.add_writer(contract_address, sender_address, address, tx_format=tx_format)


    def add_writer(self, contract_address, sender_address, address, tx_format=TxFormat.JSONRPC):
        enc = ABIContractEncoder()
        enc.method('addWriter')
        enc.typ(ABIContractType.ADDRESS)
        enc.address(address)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx


    def set_max_supply(self, contract_address, sender_address, cap, tx_format=TxFormat.JSONRPC):
        enc = ABIContractEncoder()
        enc.method('setMaxSupply')
        enc.typ(ABIContractType.UINT256)
        enc.uint256(cap)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx


    # backwards compatibility
    def remove_minter(self, contract_address, sender_address, address, tx_format=TxFormat.JSONRPC):
        return self.delete_writer(contract_address, sender_address, address, tx_format=tx_format)


    def delete_writer(self, contract_address, sender_address, address, tx_format=TxFormat.JSONRPC):
        enc = ABIContractEncoder()
        enc.method('deleteWriter')
        enc.typ(ABIContractType.ADDRESS)
        enc.address(address)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx


    def mint_to(self, contract_address, sender_address, address, value, tx_format=TxFormat.JSONRPC):
        enc = ABIContractEncoder()
        enc.method('mintTo')
        enc.typ(ABIContractType.ADDRESS)
        enc.typ(ABIContractType.UINT256)
        enc.address(address)
        enc.uint256(value)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx


    def burn(self, contract_address, sender_address, value, tx_format=TxFormat.JSONRPC):
        enc = ABIContractEncoder()
        enc.method('burn')
        enc.typ(ABIContractType.UINT256)
        enc.uint256(value)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx


    def total_burned(self, contract_address, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('totalBurned')
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def to_base_amount(self, contract_address, value, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('toBaseAmount')
        enc.typ(ABIContractType.UINT256)
        enc.uint256(value)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def remainder(self, contract_address, parts, whole, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('remainder')
        enc.typ(ABIContractType.UINT256)
        enc.typ(ABIContractType.UINT256)
        enc.uint256(parts)
        enc.uint256(whole)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def redistributions(self, contract_address, idx, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('redistributions')
        enc.typ(ABIContractType.UINT256)
        enc.uint256(idx)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def account_period(self, contract_address, address, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('accountPeriod')
        enc.typ(ABIContractType.ADDRESS)
        enc.address(address)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def to_redistribution(self, contract_address, participants, demurrage_modifier, value, period, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('toRedistribution')
        enc.typ(ABIContractType.UINT256)
        enc.typ_literal('int128')
        enc.typ(ABIContractType.UINT256)
        enc.typ(ABIContractType.UINT256)
        enc.uint256(participants)
        enc.uint256(demurrage_modifier)
        enc.uint256(value)
        enc.uint256(period)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o



    def to_redistribution_period(self, contract_address, redistribution, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('toRedistributionPeriod')
        v = strip_0x(redistribution)
        enc.typ_literal('(uint32,uint72,uint64)')
        enc.bytes32(v[:64])
        enc.bytes32(v[64:128])
        enc.bytes32(v[128:192])
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


#    def to_redistribution_participants(self, contract_address, redistribution, sender_address=ZERO_ADDRESS, id_generator=None):
#        j = JSONRPCRequest(id_generator)
#        o = j.template()
#        o['method'] = 'eth_call'
#        enc = ABIContractEncoder()
#        enc.method('toRedistributionParticipants')
#        v = strip_0x(redistribution)
#        enc.typ_literal('(uint32,uint72,uint104)')
#        #enc.typ(ABIContractType.BYTES32)
#        enc.bytes32(v[:64])
#        enc.bytes32(v[64:128])
#        enc.bytes32(v[128:192])
#        data = add_0x(enc.get())
#        tx = self.template(sender_address, contract_address)
#        tx = self.set_code(tx, data)
#        o['params'].append(self.normalize(tx))
#        o['params'].append('latest')
#        o = j.finalize(o)
#        return o
#

    def to_redistribution_supply(self, contract_address, redistribution, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('toRedistributionSupply')
        v = strip_0x(redistribution)
        enc.typ_literal('(uint32,uint72,uint64)')
        enc.bytes32(v[:64])
        enc.bytes32(v[64:128])
        enc.bytes32(v[128:192])
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def to_redistribution_demurrage_modifier(self, contract_address, redistribution, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('toRedistributionDemurrageModifier')
        v = strip_0x(redistribution)
        enc.typ_literal('(uint32,uint72,uint64)')
        enc.bytes32(v[:64])
        enc.bytes32(v[64:128])
        enc.bytes32(v[128:192])
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def base_balance_of(self, contract_address, address, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('baseBalanceOf')
        enc.typ(ABIContractType.ADDRESS)
        enc.address(address)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def set_sink_address(self, contract_address, sender_address, address, tx_format=TxFormat.JSONRPC):
        enc = ABIContractEncoder()
        enc.method('setSinkAddress')
        enc.typ(ABIContractType.ADDRESS)
        enc.address(address)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx


    def sweep(self, contract_address, sender_address, recipient_address, tx_format=TxFormat.JSONRPC):
        enc = ABIContractEncoder()
        enc.method('sweep')
        enc.typ(ABIContractType.ADDRESS)
        enc.address(recipient_address)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx



    def apply_demurrage(self, contract_address, sender_address, limit=0, tx_format=TxFormat.JSONRPC):
        if limit == 0:
            return self.transact_noarg('applyDemurrage', contract_address, sender_address)

        enc = ABIContractEncoder()
        enc.method('applyDemurrageLimited')
        enc.typ(ABIContractType.UINT256)
        enc.uint256(limit)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx


    def change_period(self, contract_address, sender_address):
        return self.transact_noarg('changePeriod', contract_address, sender_address)


    def apply_redistribution_on_account(self, contract_address, sender_address, address, tx_format=TxFormat.JSONRPC):
        enc = ABIContractEncoder()
        enc.method('applyRedistributionOnAccount')
        enc.typ(ABIContractType.ADDRESS)
        enc.address(address)
        data = enc.get()
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format)
        return tx


    def decay_level(self, contract_address, sender_address=ZERO_ADDRESS):
        return self.call_noarg('decayLevel', contract_address, sender_address=sender_address)


    def resolution_factor(self, contract_address, sender_address=ZERO_ADDRESS):
        return self.call_noarg('resolutionFactor', contract_address, sender_address=sender_address)


    def actual_period(self, contract_address, sender_address=ZERO_ADDRESS):
        return self.call_noarg('actualPeriod', contract_address, sender_address=sender_address)


    def period_start(self, contract_address, sender_address=ZERO_ADDRESS):
        return self.call_noarg('periodStart', contract_address, sender_address=sender_address)


    def period_duration(self, contract_address, sender_address=ZERO_ADDRESS):
        return self.call_noarg('periodDuration', contract_address, sender_address=sender_address)


    def demurrage_amount(self, contract_address, sender_address=ZERO_ADDRESS):
        return self.call_noarg('demurrageAmount', contract_address, sender_address=sender_address)


    def demurrage_timestamp(self, contract_address, sender_address=ZERO_ADDRESS):
        return self.call_noarg('demurrageTimestamp', contract_address, sender_address=sender_address)


    def max_supply(self, contract_address, sender_address=ZERO_ADDRESS):
        return self.call_noarg('maxSupply', contract_address, sender_address=sender_address)


#    def grow_by(self, contract_address, value, period, sender_address=ZERO_ADDRESS, id_generator=None):
#        j = JSONRPCRequest(id_generator)
#        o = j.template()
#        o['method'] = 'eth_call'
#        enc = ABIContractEncoder()
#        enc.method('growBy')
#        enc.typ(ABIContractType.UINT256)
#        enc.typ(ABIContractType.UINT256)
#        enc.uint256(value)
#        enc.uint256(period)
#        data = add_0x(enc.get())
#        tx = self.template(sender_address, contract_address)
#        tx = self.set_code(tx, data)
#        o['params'].append(self.normalize(tx))
#        o['params'].append('latest')
#        o = j.finalize(o)
#        return o
#

    def decay_by(self, contract_address, value, period, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('decayBy')
        enc.typ(ABIContractType.UINT256)
        enc.typ(ABIContractType.UINT256)
        enc.uint256(value)
        enc.uint256(period)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def get_distribution(self, contract_address, supply, demurrage_amount, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('getDistribution')
        enc.typ(ABIContractType.UINT256)
        enc.typ_literal('int128')
        enc.uint256(supply)
        enc.uint256(demurrage_amount)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def get_distribution_from_redistribution(self, contract_address, redistribution, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('getDistributionFromRedistribution')
        v = strip_0x(redistribution)
        enc.typ_literal('(uint32,uint72,uint64)')
        enc.bytes32(v[:64])
        enc.bytes32(v[64:128])
        enc.bytes32(v[128:192])
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o



    @classmethod
    def parse_actual_period(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_period_start(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_period_duration(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_demurrage_amount(self, v):
        #        return abi_decode_single(ABIContractType.UINT256, v)
        return from_fixed(v)


    @classmethod
    def parse_remainder(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_to_base_amount(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)

    
    @classmethod
    def parse_redistributions(self, v):
        return strip_0x(v)
        

    @classmethod
    def parse_account_period(self, v):
        return abi_decode_single(ABIContractType.ADDRESS, v)


    @classmethod
    def parse_to_redistribution_period(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_to_redistribution_item(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_supply_cap(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_grow_by(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_decay_by(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_get_distribution(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_decay_level(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_resolution_factor(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


    @classmethod
    def parse_total_burned(self, v):
        return abi_decode_single(ABIContractType.UINT256, v)


def bytecode(**kwargs):
    return DemurrageToken.bytecode(version=kwargs.get('version'))


def create(**kwargs):
    return DemurrageToken.cargs(
            kwargs['name'],
            kwargs['symbol'],
            kwargs['decimals'],
            kwargs['demurragelevel'],
            kwargs['redistributionperiod'],
            kwargs['sinkaddress'],
            version=kwargs.get('version'))


def args(v):
    if v == 'create':
        return (['name', 'symbol', 'decimals', 'demurragelevel', 'redistributionperiod', 'sinkaddress'], ['version'],)
    elif v == 'default' or v == 'bytecode':
        return ([], ['version'],)
    raise ValueError('unknown command: ' + v)
